tests/samplers/test_single.py

```diff
@@ -182,24 +186,74 @@ def test_iter(self, sampler: GridGeoSampler) -> None:
-        rows = int((100 - sampler.size[0]) // sampler.stride[0]) + 1
-        cols = int((100 - sampler.size[1]) // sampler.stride[1]) + 1
-        length = rows * cols * 2
+        rows = math.ceil(
+            (100 - sampler.size[0] + sampler.stride[0]) / sampler.stride[0]
+        )
+        cols = math.ceil(
+            (100 - sampler.size[1] + sampler.stride[1]) / sampler.stride[1]
+        )
+        length = rows * cols * 2  # two items in dataset
         assert len(sampler) == length
 
+    def test_len_larger(self, sampler: GridGeoSampler) -> None:
+        entire_rows = (100 - sampler.size[0] + sampler.stride[0]) // sampler.stride[0]
+        entire_cols = (100 - sampler.size[1] + sampler.stride[1]) // sampler.stride[1]
+        leftover_row = (100 - sampler.size[0] + sampler.stride[0]) / sampler.stride[
+            0
+        ] - entire_rows
+        leftover_col = (100 - sampler.size[1] + sampler.stride[1]) / sampler.stride[
+            1
+        ] - entire_cols
+        assert (
+            len(sampler)
+            == (entire_rows + math.ceil(leftover_row))
+            * (entire_cols + math.ceil(leftover_col))
+            * 2
+        )
+
     def test_roi(self, dataset: CustomGeoDataset) -> None:
         roi = BoundingBox(0, 50, 200, 250, 400, 450)
         sampler = GridGeoSampler(dataset, 2, 1, roi=roi)
         for query in sampler:
             assert query in roi
 
     def test_small_area(self) -> None:
+        ds = CustomGeoDataset()
+        ds.index.insert(0, (0, 1, 0, 1, 0, 1))
+        sampler = GridGeoSampler(ds, 2, 10)
+        assert len(sampler) == 0
+
+    # TODO: skip patches with area=0 when two tiles are
+    #  side-by-side with an overlapping edge face.
+    def test_tiles_side_by_side(self) -> None:
         ds = CustomGeoDataset()
         ds.index.insert(0, (0, 10, 0, 10, 0, 10))
-        ds.index.insert(1, (20, 21, 20, 21, 20, 21))
+        ds.index.insert(0, (0, 10, 10, 20, 0, 10))
         sampler = GridGeoSampler(ds, 2, 10)
-        for _ in sampler:
-            continue
+        for bbox in sampler:
+            assert bbox.area > 0
+
+    def test_equal_area(self) -> None:
+        ds = CustomGeoDataset()
+        ds.index.insert(0, (0, 10, 0, 10, 0, 10))
+        sampler = GridGeoSampler(ds, 10, 10, units=Units.CRS)
+        assert len(sampler) == 1
+        for bbox in sampler:
+            assert bbox == BoundingBox(
+                minx=0.0, maxx=10.0, miny=0.0, maxy=10.0, mint=0.0, maxt=10.0
+            )
+
+    def test_larger_area(self) -> None:
+        ds = CustomGeoDataset()
+        ds.index.insert(0, (0, 6, 0, 5, 0, 10))
+        sampler = GridGeoSampler(ds, 5, 5, units=Units.CRS)
+        assert len(sampler) == 2
+        assert list(sampler)[0] == BoundingBox(
+            minx=0.0, maxx=5.0, miny=0.0, maxy=5.0, mint=0.0, maxt=10.0
+        )
+        assert list(sampler)[1] == BoundingBox(
+            minx=1.0, maxx=6.0, miny=0.0, maxy=5.0, mint=0.0, maxt=10.0
+        )
```

Comment:
The point of these tests is to hardcode what the expected patches should be and how long they should be, correct? I like that idea, but still need to triple check the logic.

P.S. You can use ints instead of floats here.